packages/react-reconciler/src/ReactFiberCompleteWork.js

```diff
@@ -811,6 +813,46 @@ function completeWork(
 
         if (eventComponentInstance === null) {
           let responderState = null;
+          if (__DEV__ && !responder.allowMultipleHostChildren) {
+            let hostChildrenCount = 0;
+            let node = workInProgress.child;
+
+            mainLoop: while (node !== null) {
+              if (
+                node.tag === HostComponent ||
+                node.tag === HostText ||
+                node.tag === HostPortal
+              ) {
+                hostChildrenCount++;
+              } else {
+                const child = node.child;
+                if (child !== null) {
+                  node = child;
```

Comment:
It is not generally safe to step into a child and then follow the return pointer back because it may point to a different alternate. When we do that, we usually mutate the `return` pointer so we can get back in the right tree. Same with `sibling`.

An alternative to mutating could be to use recursion (since it's dev-only).

Reply:
This is in the complete phase though, so it shouldn't be unsafe right? You're right about recursion though, I'll change to that as it's also much easier to read.